cmd/crane/cmd/mutate.go

```diff
@@ -0,0 +1,103 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/spf13/cobra"
+)
+
+// NewCmdMutate creates a new cobra.Command for the mutate subcommand.
+func NewCmdMutate(options *[]crane.Option) *cobra.Command {
+	var lbls []string
+	var entrypoint string
+	var newRef string
+
+	mutateCmd := &cobra.Command{
+		Use:   "mutate",
+		Short: "Modify image labels and annotations",
+		Args:  cobra.ExactArgs(1),
+		Run: func(_ *cobra.Command, args []string) {
+
+			// Pull image and get config.
+			ref := args[0]
+			img, err := crane.Pull(ref, *options...)
+			if err != nil {
+				log.Fatalf("pulling %s: %v", ref, err)
+			}
+			cfg, err := img.ConfigFile()
+			if err != nil {
+				log.Fatalf("getting config: %v", err)
+			}
+			cfg = cfg.DeepCopy()
+
+			// Set labels.
+			if cfg.Config.Labels == nil {
+				cfg.Config.Labels = map[string]string{}
+			}
+			labels := map[string]string{}
+			for _, l := range lbls {
+				parts := strings.SplitN(l, "=", 2)
+				if len(parts) == 1 {
+					log.Fatalf("parsing label %q, not enough parts", l)
+				}
+				labels[parts[0]] = parts[1]
+			}
+			for k, v := range labels {
+				cfg.Config.Labels[k] = v
+			}
+
+			// Set entrypoint.
+			if entrypoint != "" {
+				cfg.Config.Entrypoint = strings.Split(entrypoint, " ")
+			}
+
+			// Mutate and write image.
+			img, err = mutate.Config(img, cfg.Config)
+			if err != nil {
+				log.Fatalf("mutating config: %v", err)
+			}
+
+			if newRef == "" {
+				newRef = ref
+			}
+			r, err := name.ParseReference(newRef)
+			if err != nil {
+				log.Fatalf("parsing %s: %v", newRef, err)
+			}
+			if _, ok := r.(name.Digest); ok {
+				newRef = r.Context().String()
+			}
+			if err := crane.Push(img, newRef, *options...); err != nil {
+				log.Fatalf("pushing %s: %v", newRef, err)
+			}
+			digest, err := img.Digest()
+			if err != nil {
+				log.Fatalf("digesting new image: %v", err)
+			}
+			fmt.Println(r.Context().Digest(digest.String()))
+		},
+	}
+	mutateCmd.Flags().StringSliceVarP(&lbls, "label", "l", nil, "New labels to add")
+	mutateCmd.Flags().StringVar(&entrypoint, "entrypoint", "", "New entrypoing to set")
+	mutateCmd.Flags().StringVar(&newRef, "new_ref", "", "New ref to apply to mutated image")
```

Comment:
I'd like to have a convention around this maybe... but I'd prefer `--tags` + `-t` for this (`ko publish` and `docker build`).

`crane append` uses `-t` and `--new_tag`.

`crane rebase` uses `--rebased`.

I think I mostly care about having the short flag `-t` work.

I guess it's possible for this to accept a digest, but that seems unlikely to be useful in practice.

This could also reasonable take a repository destination with not tags, just pushing by digest, but I don't want to complicate the UX.

Reply:
I agree. Let's also make it `crane rebase -t` in #960